compiler/backend/src/org/jetbrains/kotlin/codegen/inline/SMAP.kt

```diff
@@ -217,54 +128,40 @@ data class SMAPAndMethodNode(val node: MethodNode, val classSMAP: SMAP)
 class FileMapping(val name: String, val path: String) {
     val lineMappings = arrayListOf<RangeMapping>()
 
-    fun mapNewLineNumber(source: Int, currentIndex: Int, callSiteMarker: CallSiteMarker?): Int {
-        var mapping = lineMappings.lastOrNull()
-        if (mapping != null && mapping.callSiteMarker == callSiteMarker &&
-            (source - mapping.source) in 0 until mapping.range + (if (mapping.maxDest == currentIndex) 10 else 0)
-        ) {
-            // Save some space in the SMAP by reusing (or extending if it's the last one) the existing range.
-            mapping.range = max(mapping.range, source - mapping.source + 1)
-        } else {
-            mapping = mapNewInterval(source, currentIndex + 1, 1, callSiteMarker)
-        }
+    fun toSourceInfo(): SourceInfo =
+        SourceInfo(name, path, lineMappings.fold(0) { result, mapping -> max(result, mapping.source + mapping.range - 1) })
+
+    fun mapNewLineNumber(source: Int, currentIndex: Int, callSite: SourcePosition?): Int {
+        // Save some space in the SMAP by reusing (or extending if it's the last one) the existing range.
+        // TODO some *other* range may already cover `source`; probably too slow to check them all though.
+        //   Maybe keep the list ordered by `source` and use binary search to locate the closest range on the left?
+        val mapping = lineMappings.lastOrNull()?.takeIf { it.canReuseFor(source, currentIndex, callSite) }
+            ?: lineMappings.firstOrNull()?.takeIf { it.canReuseFor(source, currentIndex, callSite) }
```

Comment:
As I understand here you try to reuse  mapping from original source. But if it not last (previous line) then smth is already mapped and `canReuseFor` should be false, shoudn't it?

Most majority of tests have only top level declaration so file mapping mapped to class file 1:1.
Could  you add some test that contains classes so actual source range in class would be wider that actual class declaration. Would it be any problem with such cases: when finding some existing mapping outside real declaration. Could you also add navigation test for such cases?

Reply:
For every range except the last, `canReuseFor` returns true if this source line number is already within the range (and the call site marker is the same). Thus for the original file it's exactly equivalent to the old `if (sameFile && lineNumber <= smap.sourceInfo.linesInFile)` check, as `sameFile` is true if call site marker is null, and the first range is from 1 to smap.sourceInfo.linesInFile.

Reply:
I found that I've already asked you about such tests (https://github.com/JetBrains/kotlin/pull/3230#issuecomment-605899143), and they was added in #3230